pkg/io/video/copy.go

```diff
@@ -0,0 +1,196 @@
+package video
+
+import (
+	"image"
+)
+
+// Copier is an optimized image copier that tries to reduce the number of memory allocations
+// by reusing the memory area that holds pixels.
+type Copier struct {
+	buffer []uint8
+}
+
+// NewCopier creates a new copier instance and initialize internal buffer with initialSize.
+func NewCopier(initialSize int) *Copier {
+	return &Copier{
+		buffer: make([]uint8, initialSize),
+	}
+}
+
+func (copier *Copier) storeInOrder(srcs ...[]uint8) {
+	var neededSize int
+
+	for _, src := range srcs {
+		neededSize += len(src)
+	}
+
+	if len(copier.buffer) < neededSize {
+		if cap(copier.buffer) >= neededSize {
+			copier.buffer = copier.buffer[:neededSize]
+		} else {
+			copier.buffer = make([]uint8, neededSize)
+		}
+	}
+
+	var currentLen int
+	for _, src := range srcs {
+		copy(copier.buffer[currentLen:], src)
+		currentLen += len(src)
+	}
+}
+
+// Copy copies src to dst. Since dst can be nil, the caller is responsible to assign the
+// return value from Copy to dst. When dst is not nil, Copy will reuse its memory and only
+// copy values from src to dst.
+func (copier *Copier) Copy(dst, src image.Image) image.Image {
+	switch src := src.(type) {
+	case *image.Alpha:
+		clone, ok := dst.(*image.Alpha)
+		if ok {
+			*clone = *src
+		} else {
+			copied := *src
+			clone = &copied
```

Comment:
I feel it's better to return error if type of src and dst is not matched.

Reply:
Why is that? The thing is if `dst` is nil, ok will be equal to false anyway. So, I think it'll only add extra codes and more overhead.

Reply:
I think if user try to copy `*image.CMYK` into `*image.RGBA`, it's misuse of the function.

Reply:
Yep, makes sense. I think we'll just enforce this with panic since this is a misuse.

Reply:
After second thought, I'd like to keep the behaviour as it is. My motivation is that I want to use Copy for another function, "broadcast". So, it's important to keep it generic since users might change the image format at anytime in the pipeline.